File: apps/backend/src/modules/dashboard/dto/dashboard-analytics.dto.ts
import { ApiProperty } from '@nestjs/swagger';

export class StatusDistributionDto {
  @ApiProperty({ description: 'Status name' })
  status!: string;

  @ApiProperty({ description: 'Count of tickets with this status' })
  count!: number;

  @ApiProperty({ description: 'Percentage of total tickets' })
  percentage!: number;
}

export class PriorityDistributionDto {
  @ApiProperty({ description: 'Priority level' })
  priority!: string;

  @ApiProperty({ description: 'Count of tickets with this priority' })
  count!: number;

  @ApiProperty({ description: 'Percentage of total tickets' })
  percentage!: number;
}

export class DepartmentDistributionDto {
  @ApiProperty({ description: 'Department name' })
  departmentName!: string;

  @ApiProperty({ description: 'Count of tickets for this department' })
  count!: number;

  @ApiProperty({ description: 'Percentage of total tickets' })
  percentage!: number;
}

export class TrendDataPointDto {
  @ApiProperty({ description: 'Date in ISO format' })
  date!: string;

  @ApiProperty({ description: 'Number of tickets created' })
  created!: number;

  @ApiProperty({ description: 'Number of tickets resolved' })
  resolved!: number;

  @ApiProperty({ description: 'Number of tickets in progress' })
  inProgress!: number;
}

export class PerformanceMetricsDto {
  @ApiProperty({ description: 'Average resolution time in hours' })
  averageResolutionTime!: number;

  @ApiProperty({ description: 'Average first response time in hours' })
  averageFirstResponseTime!: number;

  @ApiProperty({ description: 'SLA compliance percentage' })
  slaCompliance!: number;

  @ApiProperty({ description: 'Total tickets resolved this period' })
  ticketsResolved!: number;
}

export class PredictionDto {
  @ApiProperty({ description: 'Predicted ticket volume for next period' })
  predictedVolume!: number;

  @ApiProperty({ description: 'Confidence level (0-100)' })
  confidence!: number;

  @ApiProperty({ description: 'Trend direction (increasing, decreasing, stable)' })
  trend!: string;

  @ApiProperty({ description: 'Expected resolution time in hours' })
  expectedResolutionTime!: number;
}

export class VillaDistributionDto {
  @ApiProperty({ description: 'Villa number' })
  villaNumber!: string;

  @ApiProperty({ description: 'Count of tickets for this villa' })
  count!: number;

  @ApiProperty({ description: 'Percentage of total tickets' })
  percentage!: number;
}

export class TechnicianPerformanceDto {
  @ApiProperty({ description: 'Technician ID' })
  technicianId!: string;

  @ApiProperty({ description: 'Technician name' })
  technicianName!: string;

  @ApiProperty({ description: 'Total tickets assigned' })
  totalTickets!: number;

  @ApiProperty({ description: 'Tickets completed' })
  completedTickets!: number;

  @ApiProperty({ description: 'Average resolution time in hours' })
  averageResolutionTime!: number;

  @ApiProperty({ description: 'Completion rate percentage' })
  completionRate!: number;
}

export class DashboardAnalyticsDto {
  @ApiProperty({ description: 'Status distribution', type: [StatusDistributionDto] })
  statusDistribution!: StatusDistributionDto[];

  @ApiProperty({ description: 'Priority distribution', type: [PriorityDistributionDto] })
  priorityDistribution!: PriorityDistributionDto[];

  @ApiProperty({ description: 'Department distribution', type: [DepartmentDistributionDto] })
  departmentDistribution!: DepartmentDistributionDto[];

  @ApiProperty({ description: 'Villa distribution', type: [VillaDistributionDto] })
  villaDistribution!: VillaDistributionDto[];

  @ApiProperty({ description: 'Technician performance', type: [TechnicianPerformanceDto] })
  technicianPerformance!: TechnicianPerformanceDto[];

  @ApiProperty({ description: 'Trend data over time', type: [TrendDataPointDto] })
  trends!: TrendDataPointDto[];

  @ApiProperty({ description: 'Performance metrics', type: PerformanceMetricsDto })
  performance!: PerformanceMetricsDto;

  @ApiProperty({ description: 'Predictions', type: PredictionDto })
  predictions!: PredictionDto;
}

File: apps/frontend/lib/src/features/dashboard/data/dto/dashboard_analytics_dto.dart
class StatusDistributionDto {
  StatusDistributionDto({
    required this.status,
    required this.count,
    required this.percentage,
  });

  final String status;
  final int count;
  final double percentage;

  factory StatusDistributionDto.fromJson(Map<String, dynamic> json) =>
      StatusDistributionDto(
        status: json['status'] as String,
        count: (json['count'] as num?)?.toInt() ?? 0,
        percentage: (json['percentage'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'status': status,
        'count': count,
        'percentage': percentage,
      };
}

class PriorityDistributionDto {
  PriorityDistributionDto({
    required this.priority,
    required this.count,
    required this.percentage,
  });

  final String priority;
  final int count;
  final double percentage;

  factory PriorityDistributionDto.fromJson(Map<String, dynamic> json) =>
      PriorityDistributionDto(
        priority: json['priority'] as String,
        count: (json['count'] as num?)?.toInt() ?? 0,
        percentage: (json['percentage'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'priority': priority,
        'count': count,
        'percentage': percentage,
      };
}

class DepartmentDistributionDto {
  DepartmentDistributionDto({
    required this.departmentName,
    required this.count,
    required this.percentage,
  });

  final String departmentName;
  final int count;
  final double percentage;

  factory DepartmentDistributionDto.fromJson(Map<String, dynamic> json) =>
      DepartmentDistributionDto(
        departmentName: json['departmentName'] as String,
        count: (json['count'] as num?)?.toInt() ?? 0,
        percentage: (json['percentage'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'departmentName': departmentName,
        'count': count,
        'percentage': percentage,
      };
}

class TrendDataPointDto {
  TrendDataPointDto({
    required this.date,
    required this.created,
    required this.resolved,
    required this.inProgress,
  });

  final String date;
  final int created;
  final int resolved;
  final int inProgress;

  factory TrendDataPointDto.fromJson(Map<String, dynamic> json) =>
      TrendDataPointDto(
        date: json['date'] as String,
        created: (json['created'] as num?)?.toInt() ?? 0,
        resolved: (json['resolved'] as num?)?.toInt() ?? 0,
        inProgress: (json['inProgress'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'date': date,
        'created': created,
        'resolved': resolved,
        'inProgress': inProgress,
      };
}

class PerformanceMetricsDto {
  PerformanceMetricsDto({
    required this.averageResolutionTime,
    required this.averageFirstResponseTime,
    required this.slaCompliance,
    required this.ticketsResolved,
  });

  final double averageResolutionTime;
  final double averageFirstResponseTime;
  final double slaCompliance;
  final int ticketsResolved;

  factory PerformanceMetricsDto.fromJson(Map<String, dynamic> json) =>
      PerformanceMetricsDto(
        averageResolutionTime:
            (json['averageResolutionTime'] as num?)?.toDouble() ?? 0.0,
        averageFirstResponseTime:
            (json['averageFirstResponseTime'] as num?)?.toDouble() ?? 0.0,
        slaCompliance: (json['slaCompliance'] as num?)?.toDouble() ?? 0.0,
        ticketsResolved: (json['ticketsResolved'] as num?)?.toInt() ?? 0,
      );

  Map<String, dynamic> toJson() => {
        'averageResolutionTime': averageResolutionTime,
        'averageFirstResponseTime': averageFirstResponseTime,
        'slaCompliance': slaCompliance,
        'ticketsResolved': ticketsResolved,
      };
}

class PredictionDto {
  PredictionDto({
    required this.predictedVolume,
    required this.confidence,
    required this.trend,
    required this.expectedResolutionTime,
  });

  final int predictedVolume;
  final int confidence;
  final String trend;
  final double expectedResolutionTime;

  factory PredictionDto.fromJson(Map<String, dynamic> json) => PredictionDto(
        predictedVolume: (json['predictedVolume'] as num?)?.toInt() ?? 0,
        confidence: (json['confidence'] as num?)?.toInt() ?? 0,
        trend: json['trend'] as String,
        expectedResolutionTime:
            (json['expectedResolutionTime'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'predictedVolume': predictedVolume,
        'confidence': confidence,
        'trend': trend,
        'expectedResolutionTime': expectedResolutionTime,
      };
}

class VillaDistributionDto {
  VillaDistributionDto({
    required this.villaNumber,
    required this.count,
    required this.percentage,
  });

  final String villaNumber;
  final int count;
  final double percentage;

  factory VillaDistributionDto.fromJson(Map<String, dynamic> json) =>
      VillaDistributionDto(
        villaNumber: json['villaNumber']?.toString() ?? '',
        count: (json['count'] as num?)?.toInt() ?? 0,
        percentage: (json['percentage'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'villaNumber': villaNumber,
        'count': count,
        'percentage': percentage,
      };
}

class TechnicianPerformanceDto {
  TechnicianPerformanceDto({
    required this.technicianId,
    required this.technicianName,
    required this.totalTickets,
    required this.completedTickets,
    required this.averageResolutionTime,
    required this.completionRate,
  });

  final String technicianId;
  final String technicianName;
  final int totalTickets;
  final int completedTickets;
  final double averageResolutionTime;
  final double completionRate;

  factory TechnicianPerformanceDto.fromJson(Map<String, dynamic> json) =>
      TechnicianPerformanceDto(
        technicianId: json['technicianId'] as String,
        technicianName: json['technicianName'] as String,
        totalTickets: (json['totalTickets'] as num?)?.toInt() ?? 0,
        completedTickets: (json['completedTickets'] as num?)?.toInt() ?? 0,
        averageResolutionTime:
            (json['averageResolutionTime'] as num?)?.toDouble() ?? 0.0,
        completionRate: (json['completionRate'] as num?)?.toDouble() ?? 0.0,
      );

  Map<String, dynamic> toJson() => {
        'technicianId': technicianId,
        'technicianName': technicianName,
        'totalTickets': totalTickets,
        'completedTickets': completedTickets,
        'averageResolutionTime': averageResolutionTime,
        'completionRate': completionRate,
      };
}

class DashboardAnalyticsDto {
  DashboardAnalyticsDto({
    required this.statusDistribution,
    required this.priorityDistribution,
    required this.departmentDistribution,
    required this.villaDistribution,
    required this.technicianPerformance,
    required this.trends,
    required this.performance,
    required this.predictions,
  });

  final List<StatusDistributionDto> statusDistribution;
  final List<PriorityDistributionDto> priorityDistribution;
  final List<DepartmentDistributionDto> departmentDistribution;
  final List<VillaDistributionDto> villaDistribution;
  final List<TechnicianPerformanceDto> technicianPerformance;
  final List<TrendDataPointDto> trends;
  final PerformanceMetricsDto performance;
  final PredictionDto predictions;

  factory DashboardAnalyticsDto.fromJson(Map<String, dynamic> json) {
    // Handle wrapped response format
    dynamic data = json;
    if (data is Map<String, dynamic> && data.containsKey('data')) {
      data = data['data'];
    }

    return DashboardAnalyticsDto(
      statusDistribution: (data['statusDistribution'] as List<dynamic>?)
              ?.map((e) => StatusDistributionDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      priorityDistribution: (data['priorityDistribution'] as List<dynamic>?)
              ?.map((e) => PriorityDistributionDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      departmentDistribution: (data['departmentDistribution'] as List<dynamic>?)
              ?.map((e) => DepartmentDistributionDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      villaDistribution: (data['villaDistribution'] as List<dynamic>?)
              ?.map((e) => VillaDistributionDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      technicianPerformance: (data['technicianPerformance'] as List<dynamic>?)
              ?.map((e) => TechnicianPerformanceDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      trends: (data['trends'] as List<dynamic>?)
              ?.map((e) => TrendDataPointDto.fromJson(
                    e as Map<String, dynamic>,
                  ))
              .toList() ??
          [],
      performance: PerformanceMetricsDto.fromJson(
        data['performance'] as Map<String, dynamic>,
      ),
      predictions: PredictionDto.fromJson(
        data['predictions'] as Map<String, dynamic>,
      ),
    );
  }

  Map<String, dynamic> toJson() => {
        'statusDistribution':
            statusDistribution.map((e) => e.toJson()).toList(),
        'priorityDistribution':
            priorityDistribution.map((e) => e.toJson()).toList(),
        'departmentDistribution':
            departmentDistribution.map((e) => e.toJson()).toList(),
        'villaDistribution': villaDistribution.map((e) => e.toJson()).toList(),
        'technicianPerformance':
            technicianPerformance.map((e) => e.toJson()).toList(),
        'trends': trends.map((e) => e.toJson()).toList(),
        'performance': performance.toJson(),
        'predictions': predictions.toJson(),
      };
}
